### PR TITLE
feat: implement opentelemetry

### PR DESCRIPTION
**Pull request by batrov**
- Add OpenTelemetry implementation
- Modify RecordRun and RecordPlugin to accept context
- Modify monitor as slices of monitor
- Fix statsd unit tests

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -228,7 +242,24 @@ func (r *Agent) setupExtractor(ctx context.Context, sr recipe.PluginRecipe, str
...
-func (r *Agent) setupProcessor(ctx context.Context, pr recipe.PluginRecipe, str *stream) error {
+func (r *Agent) setupProcessor(ctx context.Context, pr recipe.PluginRecipe, str *stream, recipeName string) error {
+	var (
+		pluginInfo = PluginInfo{
+			RecipeName: recipeName,
+			PluginName: pr.Name,
+			PluginType: "processor",
+			StartTime:  time.Now(),
+		}
+		err error
```
Comment (sudo-suhas) [inline agent/agent.go]: `err` is being redefined on line 263. We should instead have a named return value and check that in the deferred function instead.

File: agent/agent.go [hunk 2]
```diff
@@ -26,18 +26,28 @@ type Agent struct {
...
+	Retries    int64
+	BatchSize  int
+}
+
 // NewAgent returns an Agent with plugin factories.
 func NewAgent(config Config) *Agent {
 	mt := config.Monitor
 	if isNilMonitor(mt) {
-		mt = new(defaultMonitor)
+		mt = append(mt, new(defaultMonitor))
```
Comment (sudo-suhas) [inline agent/agent.go]: I think we can undo this change.

File: agent/agent.go [hunk 3]
```diff
@@ -228,7 +242,24 @@ func (r *Agent) setupExtractor(ctx context.Context, sr recipe.PluginRecipe, str
...
+			PluginType: "processor",
+			StartTime:  time.Now(),
+		}
+		err error
+	)
+
+	defer func() {
+		pluginInfo.Success = err == nil
+		for _, monitor := range r.monitor {
+			monitor.RecordPlugin(ctx, pluginInfo)
```
Comment (sudo-suhas) [inline agent/agent.go]: This won't work. We would end up recording the metrics before even the processor has had a chance to run. Instead we should look to implement an `OtelProcessorMiddleware` that can emit metrics on each call of `Process`.

File: agent/agent.go [hunk 4]
```diff
@@ -273,8 +312,15 @@ func (r *Agent) setupSink(ctx context.Context, sr recipe.PluginRecipe, stream *s
 			retryNotification,
 		)
 
-		success := err == nil
-		r.monitor.RecordPlugin(recipe.Name, sr.Name, "sink", success)
+		pluginInfo.Success = err == nil
+		for _, val := range stream.subscribers {
+			pluginInfo.BatchSize += val.batchSize
+		}
```
Comment (sudo-suhas) [inline agent/agent.go]: Why would the batch size be determined by the number of subscribers?

File: agent/agent.go [hunk 5]
```diff
@@ -273,8 +312,15 @@ func (r *Agent) setupSink(ctx context.Context, sr recipe.PluginRecipe, stream *s
...
+
+		for _, monitor := range r.monitor {
+			monitor.RecordPlugin(ctx, pluginInfo)
```
Comment (sudo-suhas) [inline agent/agent.go]: This won't work. We would end up recording the sink duration as a continuously increasing value since we are calculating the duration against when this function was called first and the number of retries also wouldn't be correct - we would track 1, 2, 3 retries bringing the total to 6 when the actual number of retries was 3. Instead we should look to implement an `OtelSinkMiddleware` that can emit metrics on each call of `Process`.

File: agent/monitor.go [hunk 6]
```diff
@@ -1,25 +1,24 @@
...
-func (m *defaultMonitor) RecordPlugin(recipeName, pluginName, pluginType string, success bool) {
+func (*defaultMonitor) RecordPlugin(context.Context, PluginInfo) {
 }
 
-func isNilMonitor(monitor Monitor) bool {
-	v := reflect.ValueOf(monitor)
-	return !v.IsValid() || reflect.DeepEqual(v.Interface(), reflect.Zero(v.Type()).Interface())
+func isNilMonitor(monitor []Monitor) bool {
+	return len(monitor) == 0
 }
```
Comment (sudo-suhas) [inline agent/monitor.go]: Do we need this function?

File: metrics/opentelemetry.go [hunk 7]
```diff
@@ -0,0 +1,252 @@
+package metrics
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go:1]: Can we segregate the code for Otel initialisation and the `OtelMonitor`?

**Review by sudo-suhas — CHANGES_REQUESTED** [review]

File: agent/agent.go (hunk 1 above)
Comment (batrov) [reply agent/agent.go]: updated

File: agent/agent.go (hunk 2 above)
Comment (batrov) [reply agent/agent.go]: updated

File: agent/monitor.go (hunk 6 above)
Comment (batrov) [reply agent/monitor.go]: deleted

File: metrics/opentelemetry.go (hunk 7 above)
Comment (batrov) [reply metrics/opentelemetry.go:1]: sure, updated

File: agent/agent.go (hunk 4 above)
Comment (batrov) [reply agent/agent.go]: any suggestion on how to calculate the batch size for sinks?

Comment (haveiss) [reply agent/agent.go]: cc: @spy16 @haveiss

File: agent/agent.go (hunk 3 above)
Comment (batrov) [reply agent/agent.go]: updated

File: agent/agent.go (hunk 5 above)
Comment (batrov) [reply agent/agent.go]: updated, but still got doubt on batch_size and retry count. Currently adding 1 on every RecordPlugin call

File: agent/agent.go [hunk 8]
```diff
@@ -180,6 +183,7 @@ func (r *Agent) Run(ctx context.Context, recipe recipe.Recipe) (run Run) {
 		}()
 
 		retryNotification := func(e error, d time.Duration) {
+			atomic.AddInt64(&extractorRetryCnt, 1)
```
Comment (sudo-suhas) [inline agent/agent.go:186]: We should be emitting the metric from here instead of incrementing the counter.

File: agent/agent.go [hunk 9]
```diff
@@ -267,14 +289,27 @@ func (r *Agent) setupSink(ctx context.Context, sr recipe.PluginRecipe, stream *s
...
 		err := r.retrier.retry(
 			ctx,
-			func() error { return sink.Sink(ctx, records) },
+			func() error {
+				if r.otelEnabled {
+					sink, err = otelmw.WithSinkMW(sink, sr.Name)
+					if err != nil {
+						return err
+					}
+				}
```
Comment (sudo-suhas) [inline agent/agent.go]: We should do this after line 271 and not within each retry attempt.

File: metrics/opentelemetry.go [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.20.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/encoding/gzip"
+)
+
+const gracePeriod = 5 * time.Second
+
+// OtelMonitor represents the otel monitor.
+type OtelMonitor struct {
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go]: Can we split the code for `OtelMonitor` into a different file from this one?

File: metrics/opentelemetry.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+}
+
+func InitOtel(ctx context.Context, cfg config.Config, logger *log.Logrus, appVersion string) (func(), error) {
+	res, err := resource.New(ctx,
+		resource.WithFromEnv(),
+		resource.WithTelemetrySDK(),
+		resource.WithOS(),
+		resource.WithHost(),
+		resource.WithProcess(),
+		resource.WithContainer(),
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go]: I think we can remove this, it might create too much noise.

File: metrics/opentelemetry.go [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	return res
+}
+
+// RecordRun records a run behavior
+func (m *OtelMonitor) RecordRun(ctx context.Context, run agent.Run) {
+	span := trace.SpanFromContext(ctx)
+	span.SetAttributes(attribute.String("recipe_name", run.Recipe.Name))
+
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go]:
```suggestion
```

Don't think this would be useful since we are not instrumenting traces.

File: metrics/opentelemetry.go [hunk 13]
```diff
@@ -0,0 +1,226 @@
...
+	m.recipeDuration.Record(ctx,
+		int64(run.DurationInMs),
+		metric.WithAttributes(
+			attribute.String("extractor", run.Recipe.Source.Name),
+			attribute.StringSlice("processors", getSliceStringPluginNames(run.Recipe.Processors)),
+			attribute.StringSlice("sinks", getSliceStringPluginNames(run.Recipe.Sinks)),
+			attribute.Bool("success", run.Success),
+		))
+
+	m.extractorRetries.Add(ctx, int64(run.ExtractorRetries))
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go]: Can we add the extractor name as an attribute for this?

File: metrics/opentelemetry.go [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+		int64(run.AssetsExtracted),
+		metric.WithAttributes(
+			attribute.String("extractor", run.Recipe.Source.Name),
+			attribute.StringSlice("processors", getSliceStringPluginNames(run.Recipe.Processors)),
+			attribute.StringSlice("sinks", getSliceStringPluginNames(run.Recipe.Sinks)),
+		))
+}
+
+// RecordPlugin records a individual plugin behavior in a run, this is being handled in otelmw
+func (*OtelMonitor) RecordPlugin(_ context.Context, _ agent.PluginInfo) {}
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go]:
```suggestion
func (*OtelMonitor) RecordPlugin(context.Context, agent.PluginInfo) {}
```

File: metrics/opentelemetry.go [hunk 15]
```diff
@@ -0,0 +1,226 @@
...
+
+func (m *OtelMonitor) RecordPluginRetryCount(ctx context.Context, pluginInfo agent.PluginInfo) {
```
Comment (sudo-suhas) [inline metrics/opentelemetry.go]: Should we rename this to `RecordSinkRetryCount`?

**Review by sudo-suhas — CHANGES_REQUESTED** [review]

File: metrics/opentelemetry.go (hunk 12 above)
Comment (batrov) [reply metrics/opentelemetry.go]: deleted

File: metrics/opentelemetry.go (hunk 13 above)
Comment (batrov) [reply metrics/opentelemetry.go]: updated

File: metrics/opentelemetry.go (hunk 14 above)
Comment (batrov) [reply metrics/opentelemetry.go]: updated

File: metrics/opentelemetry.go (hunk 15 above)
Comment (batrov) [reply metrics/opentelemetry.go]: we can. currently OtelMonitor are being used by plugins regardless of plugin types so thinking of this, updated.

File: metrics/opentelemetry.go (hunk 11 above)
Comment (batrov) [reply metrics/opentelemetry.go]: updated

File: metrics/opentelemetry.go (hunk 10 above)
Comment (batrov) [reply metrics/opentelemetry.go]: updated

File: agent/agent.go (hunk 9 above)
Comment (batrov) [reply agent/agent.go]: updated

**Review by sudo-suhas — APPROVED** [review]
